File: dart/container-with-most-water.dart
import 'dart:math';

class Solution {
  int maxArea(List<int> height) {
    int l = 0;
    int r = height.length - 1;
    int mostWater = 0;
    while (l < r) {
      int water = (r - l) * min(height[l], height[r]);
      mostWater = max(mostWater, water);
      if (height[l] < height[r]) {
        l++;
      } else {
        r--;
      }
    }
    return mostWater;
  }
}

File: go/container-with-most-water.go
func maxArea(height []int) (most_water int) {
	l := 0
	r := len(height) - 1
	most_water = 0
	for ; l < r; {
		water := min(height[l], height[r]) * (r - l)
		most_water = max(most_water, water)
		if height[l] < height[r] {
			l += 1
		} else {
			r -= 1
		}
	}
	return
}
